### PR TITLE
Optimize classpath search

### PR DESCRIPTION
**Pull request by dbmeneses**
The goal is to optimize the initialization of main and test classpaths, and also to get rid of old utilities to find files recursively and switch to JDK 7 NIO, with a much more readable code.
Here are the benchmark results when doing the following searches in my local mvn repo, with **4903 jars**:

- /repository
**`952ms -> 85ms`**
- /repository/
`735ms -> 84ms`
- /repository/* * /*.jar
`321ms -> 175ms`
- /repository/* *.jar
`291ms -> 168ms`
- /repository/* * /org/**.jar
`342ms -> 169ms`
- /repository/* * /* 
**`3160ms -> 220ms`**
- list of all 4903 files
`10ms -> 10ms`

Note that the syntax for patterns is now strictly GLOB, meaning that Windows path separators won't work, at least on Linux (one test removed for this).

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 1]
```diff
@@ -21,31 +21,31 @@
...
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 
-import static org.apache.commons.io.filefilter.FileFilterUtils.suffixFileFilter;
+import javax.annotation.Nullable;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MINOR Remove this unused import 'javax.annotation.Nullable'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 2]
```diff
@@ -95,79 +95,140 @@ private boolean hasJavaSources() {
     return fs.hasFiles(fs.predicates().and(fs.predicates().hasLanguage("java"), fs.predicates().hasType(fileType)));
   }
 
+  private static List<File> getFilesForPattern(Path baseDir, String pathPattern, boolean libraryProperty) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "getFilesForPattern" is 19 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 3]
```diff
@@ -95,79 +95,140 @@ private boolean hasJavaSources() {
...
+      if (Files.isDirectory(filePath)) {
+        if (libraryProperty) {
+          List<File> matches = new LibraryFinder().find(filePath, p -> true);
+          matches.add(filePath.toFile());
+          return matches;
+        } else {
+          return Collections.singletonList(filePath.toFile());
+        }
       }
+    } catch (IOException e) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MINOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by dbmeneses** [comment]
@benzonico I changed this to:

- Adapt path separator to the platform (test restored)
- Calculate internally a Set to avoid dups, but exposes in the end a List because I don't want to change all the classes using the classpath (can be done later). Performance hit in the worst case tested was from `220ms -> 231ms`.

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 4]
```diff
@@ -95,110 +96,149 @@ private boolean hasJavaSources() {
     return fs.hasFiles(fs.predicates().and(fs.predicates().hasLanguage("java"), fs.predicates().hasType(fileType)));
   }
 
+  private static Set<File> getFilesForPattern(Path baseDir, String pathPattern, boolean libraryProperty) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "getFilesForPattern" is 19 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 5]
```diff
@@ -95,110 +96,149 @@ private boolean hasJavaSources() {
...
+      if (Files.isDirectory(filePath)) {
+        if (libraryProperty) {
+          Set<File> matches = new LibraryFinder().find(filePath, p -> true);
+          matches.add(filePath.toFile());
+          return matches;
+        } else {
+          return Collections.singleton(filePath.toFile());
+        }
+      }
+    } catch (IOException e) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MINOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by dbmeneses** [comment]
Added the special case for target/classes and target/test-classes with a unit test.

File: java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java [hunk 6]
```diff
@@ -95,110 +97,159 @@ private boolean hasJavaSources() {
...
+    try {
+      Path filePath = resolvePath(baseDir, pathPattern);
+
+      if ((pathPattern.endsWith(".jar") || pathPattern.endsWith(".zip")) && Files.isRegularFile(filePath)) {
+        return Collections.singleton(filePath.toFile());
+      }
+      if (Files.isDirectory(filePath)) {
+        return getMatchesInDir(filePath, libraryProperty);
       }
+    } catch (IOException e) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/AbstractJavaClasspath.java]: MINOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**